core/lib/contract_verifier/src/etherscan/mod.rs

```diff
@@ -0,0 +1,459 @@
+        tracing::warn!("Failed to process etherscan verification request with id = {job_id}");
+        let mut connection = self
+            .connection_pool
+            .connection_tagged("etherscan_verifier")
+            .await?;
+
+        connection
+            .etherscan_verification_dal()
+            .save_verification_result(
+                job_id,
+                EtherscanVerificationJobResultStatus::Failed,
+                Some(&error),
+            )
+            .await?;
+
+        API_CONTRACT_VERIFIER_METRICS.failed_verifications[&"etherscan_verifier"].inc();
+
+        Ok(())
+    }
+
+    async fn reset_etherscan_verification_processing_started_at(
+        &self,
+        request_id: usize,
+    ) -> Result<(), VerifierError> {
+        let mut connection = self
+            .connection_pool
+            .connection_tagged("etherscan_verifier")
+            .await?;
+
+        connection
+            .etherscan_verification_dal()
+            .reset_processing_started_at(request_id)
+            .await?;
+
+        Ok(())
+    }
+
+    async fn send_verification_request(
+        &self,
+        verification_request: &VerificationRequest,
+    ) -> Result<Option<String>, VerifierError> {
+        let req = verification_request.req.clone();
+        tracing::info!(
+            "Sending verification request to Etherscan, address = {:#?}",
+            req.contract_address,
+        );
+        let request_id = match self
+            .client
+            .verify(EtherscanVerificationRequest::from_verification_request(
+                req.clone(),
+                &self.solc_versions_fetcher,
+            ))
+            .await
+        {
+            Ok(id) => id,
+            // Even though there is a call to check if the contract is already verified
+            // we still need to process ContractAlreadyVerified response from the verification
+            // API call. This is because the contract can be verified by another party.
+            Err(EtherscanError::ContractAlreadyVerified) => {
+                tracing::info!(
+                    "Contract with address {:#?} is already verified",
+                    req.contract_address
+                );
+                return Ok(None);
+            }
+            Err(e) => return Err(e.into()),
+        };
+        let mut connection = self
+            .connection_pool
+            .connection_tagged("etherscan_verifier")
+            .await?;
+
+        connection
+            .etherscan_verification_dal()
+            .save_etherscan_verification_id(verification_request.id, &request_id)
+            .await?;
+
+        Ok(Some(request_id))
+    }
+
+    async fn await_verification(
+        &mut self,
+        verification_request: &VerificationRequest,
+        verification_id: String,
+    ) -> Result<(), VerifierError> {
+        let mut get_status_attempts = 0;
+        loop {
+            get_status_attempts += 1;
```

Comment:
Nit: you can make this loop immutable if you will change the default return value to `Err`, e.g. (pseudocode):

```rust
for attempt in 1..=MAX_POLL_VERIFICATION_RESULT_ATTEMPTS {
  match self.client.get_verification_status(&verification_id).await {
    Ok(_) => return Ok(()),
    Err(EtherscanError::VerificationPending) => {
      if attempt < MAX_POLL_VERIFICATION_RESULT_ATTEMPTS {
        self.sleep(POLL_VERIFICATION_RESULT_INTERVAL).await?;
      }
    }
    Err(e) => return Err(e.into()),
  }
}

Err(ProcessingError::VerificationStatusPollingTimeout.into())
```